src/actors/builtin/storage_miner/storage_miner_actor_code.go

```diff
@@ -23,11 +24,43 @@ import (
 // Actor methods
 ////////////////////////////////////////////////////////////////////////////////
 
-func (a *StorageMinerActorCode_I) GetWorkerVRFKey(rt Runtime) filcrypto.VRFPublicKey {
+func (a *StorageMinerActorCode_I) StageWorkerKeyChange(rt Runtime, key filcrypto.VRFPublicKey) (err error) {
 	h, st := a.State(rt)
-	ret := st.Info().WorkerVRFKey()
+	rt.ValidateImmediateCallerIs(st.Info().Owner())
+	util.IMPL_TODO() // verify public key syntax
+
+	keyAddr, err := addr.Address_Make_Key(node_base.NETWORK, addr.KeyHash(key))
+	if err != nil {
+		Release(rt, h, st)
+		return err
+	}
+
+	keyChange := &MinerKeyChange_I{
+		NewKey_:      keyAddr,
+		EffectiveAt_: rt.CurrEpoch() + node_base.MINER_KEY_CHANGE_FREEZE,
+	}
+
+	// note that this may replace another pending key change
+	st.Info()._updateMinerPendingKeyChange(keyChange)
 	Release(rt, h, st)
-	return ret
+	return nil
+}
+
+func (a *StorageMinerActorCode_I) CommitWorkerKeyChange(rt Runtime) (err error) {
+	h, st := a.State(rt)
+
+	if st.Info().PendingKeyChange() == nil {
+		return errors.New("No pending key change.")
+	}
+
+	// lazy updates to worker keys if at effectiveness
+	if st.Info().PendingKeyChange().EffectiveAt() <= rt.CurrEpoch() {
+		st.Info()._updateMinerKey()
+		util.TODO()
+		// remains to update the key in every deal, or associate deals to minerActorID instead of workerKey
```

Comment:
IMO deals should be associated with actors and the key looked up any time it is needed. cc @hannahhoward

Reply:
i will admit being a bit out of context here, but if the question is do we need to be able to, at any time, query the chain to determine "what deals are currently on chain for miner" the answer is yes.

Reply:
or query the actor... sorry again, out of context and my area of expertise